rfcs/0000-address-format/0000-address-format.md

```diff
@@ -0,0 +1,98 @@
+---
+Number: 21
+Category: Standards Track
+Status: Proposal
+Author: Cipher Wang
+Organization: Nervos Foundation
+Created: 2019-01-20
+---
+
+# Address Format
+
+## Abstract
+
+*CKB Address Format* is an application level cell **lock script** display recommendation. The lock script consists of three key parameters, including *code_hash*, *hash_type* and *args*. CKB address packages lock script into a single line, verifiable, and human read friendly format.
+
+## Data Structure
+
+### Payload Format Types
+
+To generate a CKB address, we firstly encode lock script to bytes array, name *payload*. And secondly, we wrap the payload into final address format.
+
+There are several methods to convert lock script into payload bytes array. We use 1 byte to identify the payload format.
+
+| format type |                   description                  |
+|:-----------:|------------------------------------------------|
+|  0x01       | short version for locks with popular code_hash |
+|  0x02       | full version with hash_type = "Data"           |
+|  0x04       | full version with hash_type = "Type"           |
+
+### Short Payload Format
+
+Short payload format is a compact format which identifies common used code_hash by 1 byte code_hash_index instead of 32 bytes code_hash.
+
+```c
+payload = 0x01 | code_hash_index | single_arg
+```
```

Comment:
We can save 1 byte here by reuse format byte to indicate `code_hash_index`. Proposal:

- use the highest bit of format byte to indicate if it's in short format or full format: 0 for short format, 1 for full format
- for short format, use the remaining bits as `code_hash_index`
- for full format, use the second bit to indicate the hash_type


| format type |                   description                  |
|:-----------:|------------------------------------------------|
|  0x00 - 0x7f       | `code_hash_index` value for short payload format |
|  0x80       | full version with hash_type = "Data"           |
|  0xc0       | full version with hash_type = "Type"           |

Reply:
It's ok, but makes incompatible with previous address format. Since we have several tools to generate addresses, explorer to show them, and users including miners are using them, I suggest not change it.

Reply:
Agree. That also makes the format a bit complicated.